File: 09_hull_white_short_rate_model/python/hull_white_simulation.py
"""
hull_white_simulation.py

Monte Carlo simulation for the Hull-White short-rate model.
"""

import numpy as np


def simulate_short_rate_paths(
    model,
    r0,
    maturity,
    n_steps,
    n_paths,
    seed=42
):
    """
    Simulate Hull-White short-rate paths using Euler discretisation.

    Parameters
    ----------
    model : HullWhiteModel
        Hull-White model.

    r0 : float
        Initial short rate.

    maturity : float
        Simulation horizon.

    n_steps : int
        Number of time steps.

    n_paths : int
        Number of Monte Carlo paths.

    seed : int
        Random seed.

    Returns
    -------
    times : np.ndarray
        Simulation time grid.

    paths : np.ndarray
        Short-rate paths with shape:

            (n_paths, n_steps + 1)
    """

    np.random.seed(seed)

    dt = maturity / n_steps

    times = np.linspace(
        0.0,
        maturity,
        n_steps + 1
    )

    paths = np.zeros(
        (
            n_paths,
            n_steps + 1
        )
    )

    paths[:, 0] = r0

    for step in range(n_steps):

        t = times[step]

        z = np.random.normal(
            size=n_paths
        )

        for path in range(n_paths):

            paths[path, step + 1] = (
                model.evolve_short_rate(
                    t,
                    paths[path, step],
                    dt,
                    z[path]
                )
            )

    return times, paths


def path_discount_factors(
    times,
    short_rate_paths
):
    """
    Compute pathwise discount factors.

    Discount factor along a path:

        D(0,T) = exp(- integral_0^T r_t dt)

    The integral is approximated using a simple left Riemann sum.

    Parameters
    ----------
    times : np.ndarray
        Simulation time grid.

    short_rate_paths : np.ndarray
        Simulated short-rate paths.

    Returns
    -------
    np.ndarray
        Pathwise discount factors to final maturity.
    """

    dt = times[1] - times[0]

    integrals = np.sum(
        short_rate_paths[:, :-1],
        axis=1
    ) * dt

    return np.exp(
        -integrals
    )


def monte_carlo_zero_coupon_bond_price(
    times,
    short_rate_paths
):
    """
    Estimate zero-coupon bond price by Monte Carlo.

    Parameters
    ----------
    times : np.ndarray
        Simulation time grid.

    short_rate_paths : np.ndarray
        Short-rate paths.

    Returns
    -------
    float
        Monte Carlo estimate of discount bond price.
    """

    dfs = path_discount_factors(
        times,
        short_rate_paths
    )

    return float(
        np.mean(dfs)
    )

File: 10_short_rate_models_with_stochastic_volatility/cpp/HullWhiteSimulation.cpp
#include "HullWhiteSimulation.hpp"

#include <random>
#include <cmath>

/*
    Simulate Hull-White short-rate paths using Euler discretisation.
*/

SimulationResult simulateShortRatePaths(
    const HullWhiteModel& model,
    double r0,
    double maturity,
    int nSteps,
    int nPaths,
    unsigned int seed
){

    double dt =
        maturity
        /
        static_cast<double>(nSteps);

    SimulationResult result;

    result.times.resize(
        nSteps + 1
    );

    for(int i = 0; i <= nSteps; ++i){

        result.times[i] =
            i * dt;
    }

    result.paths =
        std::vector<std::vector<double>>(
            nPaths,
            std::vector<double>(
                nSteps + 1,
                0.0
            )
        );

    for(int p = 0; p < nPaths; ++p){

        result.paths[p][0] =
            r0;
    }

    std::mt19937 generator(seed);

    std::normal_distribution<double> normal(
        0.0,
        1.0
    );

    for(int step = 0; step < nSteps; ++step){

        double t =
            result.times[step];

        for(int p = 0; p < nPaths; ++p){

            double z =
                normal(generator);

            result.paths[p][step + 1] =
                model.evolveShortRate(
                    t,
                    result.paths[p][step],
                    dt,
                    z
                );
        }
    }

    return result;
}

/*
    Pathwise discount factor:

        D(0,T) = exp(- integral r_t dt)
*/

double pathDiscountFactor(
    const std::vector<double>& times,
    const std::vector<double>& path
){

    double dt =
        times[1]
        -
        times[0];

    double integral =
        0.0;

    for(size_t i = 0; i < path.size() - 1; ++i){

        integral +=
            path[i]
            *
            dt;
    }

    return
        std::exp(
            -integral
        );
}

/*
    Monte Carlo zero-coupon bond price.
*/

double monteCarloZeroCouponBondPrice(
    const SimulationResult& result
){

    double sum =
        0.0;

    for(const auto& path : result.paths){

        sum +=
            pathDiscountFactor(
                result.times,
                path
            );
    }

    return
        sum
        /
        static_cast<double>(
            result.paths.size()
        );
}
